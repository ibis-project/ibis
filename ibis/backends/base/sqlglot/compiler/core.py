"""ClickHouse ibis expression to sqlglot compiler.

The compiler is built with a few `singledispatch` functions:

    1. `translate` for table expressions
    1. `translate` for table nodes
    1. `translate_rel`
    1. `translate_val`

## `translate`

### Expression Implementation

The table expression implementation of `translate` is a pass through to the
node implementation.

### Node Implementation

There's a single `ops.Node` implementation for `ops.TableNode`s instances.

This function:

    1. Topologically sorts the expression graph.
    1. Seeds the compilation cache with in-degree-zero table names.
    1. Iterates though nodes with at least one in-degree and places the result
       in the compilation cache. The cache is used to construct `ops.TableNode`
       keyword arguments to the current translation rule.

## `translate_rel`

Translates a table operation given already-translated table inputs.

If a table node needs to translate value expressions, for example, an
`ops.Aggregation` that rule is responsible for calling `translate_val`.

## `translate_val`

Recurses top-down and translates the arguments of the value expression and uses
those as input to construct the output.
"""

from __future__ import annotations

import itertools
from typing import TYPE_CHECKING, Any

import sqlglot as sg

import ibis.expr.operations as ops
import ibis.expr.types as ir
from ibis.common.deferred import _
from ibis.expr.analysis import c, p, x

if TYPE_CHECKING:
    from collections.abc import Mapping


def translate(
    op, *, params: Mapping[ir.Value, Any], translate_rel, translate_val
) -> sg.exp.Expression:
    """Translate an ibis operation to a sqlglot expression.

    Parameters
    ----------
    op
        An ibis operation
    params
        A mapping of expressions to concrete values
    translate_rel
        Relation node translator
    translate_val
        Value node translator

    Returns
    -------
    sqlglot.expressions.Expression
        A sqlglot expression
    """

    def _translate_node(node, **kwargs):
        if isinstance(node, ops.Value):
            return translate_val(node, **kwargs)
        return translate_rel(node, **kwargs)

    gen_alias_index = itertools.count()

    def fn(node, _, **kwargs):
        result = _translate_node(node, **kwargs)

        # don't alias root nodes or value ops
        if node is op or isinstance(node, ops.Value):
            return result

        alias_index = next(gen_alias_index)
        alias = f"t{alias_index:d}"

        try:
            return result.subquery(alias)
        except AttributeError:
            return sg.alias(result, alias)

    # substitute parameters immediately to avoid having to define a
    # ScalarParameter translation rule
    #
    # this lets us avoid threading `params` through every `translate_val` call
    # only to be used in the one place it would be needed: the ScalarParameter
    # `translate_val` rule
    params = {param.op(): value for param, value in params.items()}
    replace_literals = p.ScalarParameter(dtype=x) >> (
        lambda _, x: ops.Literal(value=params[_], dtype=x)
    )

    # replace any checks against an empty right side of the IN operation with
    # `False`
    replace_empty_in_values_with_false = p.InValues(..., ()) >> c.Literal(
        False, dtype="bool"
    )

    # add an ORDER BY clause to rank window functions that don't have one
    add_order_by_to_empty_window_functions = p.WindowFunction(
        p.PercentRank(x) | p.RankBase(x) | p.CumeDist(x) | p.NTile(x),
        p.WindowFrame(..., order_by=()) >> _.copy(order_by=(x,)),
    )

    # subtract one from one-based functions to convert to zero-based indexing
    subtract_one_from_one_indexed_functions = (
        p.WindowFunction(p.RankBase | p.NTile)
        | p.StringFind
        | p.FindInSet
        | p.ArrayPosition
    ) >> c.Subtract(_, 1)

    add_one_to_nth_value_input = p.NthValue >> _.copy(nth=c.Add(_.nth, 1))

    op = op.replace(
        replace_literals
        | replace_empty_in_values_with_false
        | add_order_by_to_empty_window_functions
        | subtract_one_from_one_indexed_functions
        | add_one_to_nth_value_input
    )
    # apply translate rules in topological order
    results = op.map(fn)
    node = results[op]
    return node.this if isinstance(node, sg.exp.Subquery) else node
