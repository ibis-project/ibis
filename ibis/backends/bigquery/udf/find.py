from __future__ import annotations

import ast

import toolz


class NameFinder:
    """Helper class to find the unique names in an AST."""

    __slots__ = ()

    def find(self, node):
        typename = type(node).__name__
        method = getattr(self, f"find_{typename}", None)
        if method is None:
            fields = getattr(node, "_fields", None)
            if fields is None:
                return
            for field in fields:
                value = getattr(node, field)
                yield from self.find(value)
        else:
            yield from method(node)

    def find_Name(self, node):
        # TODO not sure if this is robust to scope changes
        yield node

    def find_list(self, node):
        return list(toolz.concat(map(self.find, node)))

    def find_Call(self, node):
        if not isinstance(node.func, ast.Name):
            fields = node._fields
        else:
            fields = [field for field in node._fields if field != "func"]
        return toolz.concat(map(self.find, (getattr(node, field) for field in fields)))


def find_names(node: ast.AST) -> list[ast.Name]:
    """Return the unique `ast.Name` instances in an AST.

    Examples
    --------
    >>> import ast
    >>> node = ast.parse("a + b")
    >>> names = find_names(node)
    >>> names
    [<....Name object at 0x...>, <....Name object at 0x...>]
    >>> names[0].id
    'a'
    >>> names[1].id
    'b'
    """
    return list(
        toolz.unique(
            filter(None, NameFinder().find(node)),
            key=lambda node: (node.id, type(node.ctx)),
        )
    )
