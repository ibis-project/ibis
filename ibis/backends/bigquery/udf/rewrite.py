from __future__ import annotations

import ast
from typing import Callable


def matches(value: ast.AST, pattern: ast.AST) -> bool:
    """Check whether `value` matches `pattern`."""
    # types must match exactly
    if type(value) != type(pattern):
        return False

    # primitive value, such as None, True, False etc
    if not isinstance(value, ast.AST) and not isinstance(pattern, ast.AST):
        return value == pattern

    fields = [
        (field, getattr(pattern, field))
        for field in pattern._fields
        if hasattr(pattern, field)
    ]
    return all(
        matches(getattr(value, field_name), field_value)
        for field_name, field_value in fields
    )


class Rewriter:
    """AST pattern matcher to enable rewrite rules."""

    def __init__(self):
        self.funcs: list[tuple[ast.AST, Callable[[ast.expr], ast.expr]]] = []

    def register(self, pattern):
        def wrapper(f):
            self.funcs.append((pattern, f))
            return f

        return wrapper

    def __call__(self, node):
        # TODO: more efficient way of doing this?
        for pattern, func in self.funcs:
            if matches(node, pattern):
                return func(node)
        return node


rewrite = Rewriter()
