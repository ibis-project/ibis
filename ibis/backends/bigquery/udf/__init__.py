from __future__ import annotations

import collections
import inspect
import itertools
from typing import TYPE_CHECKING, Callable, Literal

import ibis.expr.datatypes as dt
import ibis.expr.rules as rlz
from ibis.backends.bigquery.datatypes import BigQueryType, spread_type
from ibis.backends.bigquery.operations import BigQueryUDFNode
from ibis.backends.bigquery.udf.core import PythonToJavaScriptTranslator
from ibis.legacy.udf.validate import validate_output_type

if TYPE_CHECKING:
    from collections.abc import Iterable, Mapping

__all__ = ("udf",)

_udf_name_cache: dict[str, Iterable[int]] = collections.defaultdict(itertools.count)


def _create_udf_node(name, fields):
    """Create a new UDF node type.

    Parameters
    ----------
    name : str
        Then name of the UDF node
    fields : OrderedDict
        Mapping of class member name to definition

    Returns
    -------
    result : type
        A new BigQueryUDFNode subclass
    """
    definition = next(_udf_name_cache[name])
    external_name = f"{name}_{definition:d}"
    return type(external_name, (BigQueryUDFNode,), fields)


class _BigQueryUDF:
    def __call__(self, *args, **kwargs):
        return self.python(*args, **kwargs)

    def python(
        self,
        input_type: Iterable[dt.DataType],
        output_type: dt.DataType,
        strict: bool = True,
        libraries: Iterable[str] | None = None,
        determinism: bool | None = None,
    ) -> Callable:
        '''Define a UDF for BigQuery.

        The function is transpiled to JS.

        `INT64` is not supported as an argument type or a return type, as per
        [the BigQuery documentation](https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions#sql-type-encodings-in-javascript).

        Parameters
        ----------
        input_type
            Iterable of types, one per argument.
        output_type
            Return type of the UDF.
        strict
            Whether or not to put a ``'use strict';`` string at the beginning of
            the UDF. Setting to ``False`` is probably a bad idea.
        libraries
            An iterable of Google Cloud Storage URIs containing to JavaScript source
            code. Note that any symbols (functions, classes, variables, etc.) that
            are exposed in these JavaScript files will be visible inside the UDF.
        determinism
            Provides a hint to BigQuery as to whether the query result can be cached.

        Returns
        -------
        Callable
            The wrapped user-defined function.

        Examples
        --------
        >>> from ibis.backends.bigquery import udf
        >>> import ibis.expr.datatypes as dt
        >>> @udf.python(input_type=[dt.double], output_type=dt.double)
        ... def add_one(x):
        ...     return x + 1
        ...
        >>> print(add_one.sql)
        CREATE TEMPORARY FUNCTION add_one_0(x FLOAT64)
        RETURNS FLOAT64
        LANGUAGE js AS """
        'use strict';
        function add_one(x) {
            return (x + 1);
        }
        return add_one(x);
        """;
        >>> @udf.python(
        ...     input_type=[dt.double, dt.double], output_type=dt.Array(dt.double)
        ... )
        ... def my_range(start, stop):
        ...     def gen(start, stop):
        ...         curr = start
        ...         while curr < stop:
        ...             yield curr
        ...             curr += 1
        ...
        ...     result = []
        ...     for value in gen(start, stop):
        ...         result.append(value)
        ...     return result
        >>> print(my_range.sql)
        CREATE TEMPORARY FUNCTION my_range_0(start FLOAT64, stop FLOAT64)
        RETURNS ARRAY<FLOAT64>
        LANGUAGE js AS """
        'use strict';
        function my_range(start, stop) {
            function* gen(start, stop) {
                let curr = start;
                while ((curr < stop)) {
                    yield curr;
                    curr += 1;
                }
            }
            let result = [];
            for (let value of gen(start, stop)) {
                result.push(value);
            }
            return result;
        }
        return my_range(start, stop);
        """;
        >>> @udf.python(
        ...     input_type=[dt.double, dt.double],
        ...     output_type=dt.Struct.from_tuples(
        ...         [("width", "double"), ("height", "double")]
        ...     ),
        ... )
        ... def my_rectangle(width, height):
        ...     class Rectangle:
        ...         def __init__(self, width, height):
        ...             self.width = width
        ...             self.height = height
        ...
        ...         @property
        ...         def area(self):
        ...             return self.width * self.height
        ...
        ...         def perimeter(self):
        ...             return 2 * (self.width + self.height)
        ...
        ...     return Rectangle(width, height)
        >>> print(my_rectangle.sql)
        CREATE TEMPORARY FUNCTION my_rectangle_0(width FLOAT64, height FLOAT64)
        RETURNS STRUCT<width FLOAT64, height FLOAT64>
        LANGUAGE js AS """
        'use strict';
        function my_rectangle(width, height) {
            class Rectangle {
                constructor(width, height) {
                    this.width = width;
                    this.height = height;
                }
                get area() {
                    return (this.width * this.height);
                }
                perimeter() {
                    return (2 * (this.width + this.height));
                }
            }
            return (new Rectangle(width, height));
        }
        return my_rectangle(width, height);
        """;
        '''
        validate_output_type(output_type)

        if libraries is None:
            libraries = []

        def wrapper(f):
            if not callable(f):
                raise TypeError(f"f must be callable, got {f}")

            signature = inspect.signature(f)
            parameter_names = signature.parameters.keys()
            source = PythonToJavaScriptTranslator(f).compile()
            args = ", ".join(parameter_names)
            strict_str = repr("use strict") + ";\n" if strict else ""
            function_body = f"""\
{strict_str}{source}
return {f.__name__}({args});\
"""

            return self.js(
                name=f.__name__,
                params=(dict(zip(parameter_names, input_type))),
                output_type=output_type,
                body=function_body,
                libraries=libraries,
                determinism=determinism,
            )

        return wrapper

    @staticmethod
    def js(
        name: str,
        params: Mapping[str, dt.DataType],
        output_type: dt.DataType,
        body: str,
        libraries: Iterable[str] | None = None,
        determinism: bool | None = None,
    ) -> Callable:
        '''Define a Javascript UDF for BigQuery.

        `INT64` is not supported as an argument type or a return type, as per
        [the BigQuery documentation](https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions#sql-type-encodings-in-javascript).

        Parameters
        ----------
        name:
            The name of the function.
        params
            Mapping of names and types of parameters
        output_type
            Return type of the UDF.
        body:
            The code of the function.
        libraries
            An iterable of Google Cloud Storage URIs containing to JavaScript source
            code. Note that any symbols (functions, classes, variables, etc.) that
            are exposed in these JavaScript files will be visible inside the UDF.
        determinism
            Provides a hint to BigQuery as to whether the query result can be cached.

        Returns
        -------
        Callable
            The user-defined function.

        Examples
        --------
        >>> from ibis.backends.bigquery import udf
        >>> import ibis.expr.datatypes as dt
        >>> add_one = udf.js(
        ...     name="add_one",
        ...     params={"a": dt.double},
        ...     output_type=dt.double,
        ...     body="return x + 1",
        ... )
        >>> print(add_one.sql)
        CREATE TEMPORARY FUNCTION add_one_0(x FLOAT64)
        RETURNS FLOAT64
        LANGUAGE js AS """
        return x + 1
        """;
        '''
        validate_output_type(output_type)
        if any(
            type_ == dt.int64
            for param_type in params.values()
            for type_ in spread_type(param_type)
        ) or any(type_ == dt.int64 for type_ in spread_type(output_type)):
            raise TypeError(
                "BigQuery does not support INT64 as an argument type or a return type "
                "for UDFs. Replace INT64 with FLOAT64 in your UDF signature and "
                "cast all INT64 inputs to FLOAT64."
            )

        if libraries is None:
            libraries = []

        udf_node_fields = {
            name: rlz.ValueOf(None if type_ == "ANY TYPE" else type_)
            for name, type_ in params.items()
        }

        udf_node_fields["dtype"] = output_type
        udf_node_fields["shape"] = rlz.shape_like("args")
        udf_node_fields["__slots__"] = ("sql",)

        udf_node = _create_udf_node(name, udf_node_fields)

        from ibis.backends.bigquery.compiler import compiles

        @compiles(udf_node)
        def compiles_udf_node(t, op):
            args = ", ".join(map(t.translate, op.args))
            return f"{udf_node.__name__}({args})"

        bigquery_signature = ", ".join(
            f"{name} {BigQueryType.from_ibis(dt.dtype(type_))}"
            for name, type_ in params.items()
        )
        return_type = BigQueryType.from_ibis(dt.dtype(output_type))
        libraries_opts = (
            f"\nOPTIONS (\n    library={list(libraries)!r}\n)" if libraries else ""
        )
        determinism_formatted = {
            True: "DETERMINISTIC\n",
            False: "NOT DETERMINISTIC\n",
            None: "",
        }.get(determinism)
        sql_code = f'''\
CREATE TEMPORARY FUNCTION {udf_node.__name__}({bigquery_signature})
RETURNS {return_type}
{determinism_formatted}LANGUAGE js AS """
{body}
"""{libraries_opts};'''

        def wrapped(*args, **kwargs):
            node = udf_node(*args, **kwargs)
            object.__setattr__(node, "sql", sql_code)
            return node.to_expr()

        wrapped.__signature__ = inspect.Signature(
            parameters=[
                inspect.Parameter(
                    name=param, kind=inspect.Parameter.POSITIONAL_OR_KEYWORD
                )
                for param in params.keys()
            ]
        )
        wrapped.__name__ = name
        wrapped.sql = sql_code
        return wrapped

    @staticmethod
    def sql(
        name: str,
        params: Mapping[str, dt.DataType | Literal["ANY TYPE"]],
        output_type: dt.DataType,
        sql_expression: str,
    ) -> Callable:
        """Define a SQL UDF for BigQuery.

        Parameters
        ----------
        name:
            The name of the function.
        params
            Mapping of names and types of parameters
        output_type
            Return type of the UDF.
        sql_expression
            The SQL expression that defines the function.

        Returns
        -------
        Callable
            The wrapped user-defined function.

        Examples
        --------
        >>> from ibis.backends.bigquery import udf
        >>> import ibis.expr.datatypes as dt
        >>> add_one = udf.sql(
        ...     name="add_one",
        ...     params={"x": dt.double},
        ...     output_type=dt.double,
        ...     sql_expression="x + 1",
        ... )
        >>> print(add_one.sql)
        CREATE TEMPORARY FUNCTION add_one_0(x FLOAT64)
        RETURNS FLOAT64
        AS (x + 1)
        """
        validate_output_type(output_type)
        udf_node_fields = {
            name: rlz.ValueOf(None if type_ == "ANY TYPE" else type_)
            for name, type_ in params.items()
        }
        return_type = BigQueryType.from_ibis(dt.dtype(output_type))

        udf_node_fields["dtype"] = output_type
        udf_node_fields["shape"] = rlz.shape_like("args")
        udf_node_fields["__slots__"] = ("sql",)

        udf_node = _create_udf_node(name, udf_node_fields)

        from ibis.backends.bigquery.compiler import compiles

        @compiles(udf_node)
        def compiles_udf_node(t, op):
            args_formatted = ", ".join(map(t.translate, op.args))
            return f"{udf_node.__name__}({args_formatted})"

        bigquery_signature = ", ".join(
            "{name} {type}".format(
                name=name,
                type="ANY TYPE"
                if type_ == "ANY TYPE"
                else BigQueryType.from_ibis(dt.dtype(type_)),
            )
            for name, type_ in params.items()
        )
        sql_code = f"""\
CREATE TEMPORARY FUNCTION {udf_node.__name__}({bigquery_signature})
RETURNS {return_type}
AS ({sql_expression});"""

        def wrapper(*args, **kwargs):
            node = udf_node(*args, **kwargs)
            object.__setattr__(node, "sql", sql_code)
            return node.to_expr()

        return wrapper


udf = _BigQueryUDF()
