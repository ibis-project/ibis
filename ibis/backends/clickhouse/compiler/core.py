"""ClickHouse ibis expression to sqlglot compiler.

The compiler is built with a few `singledispatch` functions:

    1. `translate_rel` for compiling `ops.TableNode`s
    1. `translate_val` for compiling `ops.Value`s

## `translate`

### Node Implementation

There's a single `ops.Node` implementation for `ops.TableNode`s instances.

This function compiles each node in topological order. The topological sorting,
result caching, and iteration are all handled by
`ibis.expr.operations.core.Node.map`.
"""

from __future__ import annotations

import itertools
from typing import TYPE_CHECKING, Any

import sqlglot as sg

import ibis.expr.operations as ops
import ibis.expr.types as ir
from ibis.backends.clickhouse.compiler.relations import translate_rel
from ibis.backends.clickhouse.compiler.values import translate_val
from ibis.common.deferred import _
from ibis.expr.analysis import c, find_first_base_table, p, x, y

if TYPE_CHECKING:
    from collections.abc import Mapping


def _translate_node(node, **kwargs):
    if isinstance(node, ops.Value):
        return translate_val(node, **kwargs)
    assert isinstance(node, ops.TableNode)
    return translate_rel(node, **kwargs)


def translate(op: ops.TableNode, params: Mapping[ir.Value, Any]) -> sg.exp.Expression:
    """Translate an ibis operation to a sqlglot expression.

    Parameters
    ----------
    op
        An ibis `TableNode`
    params
        A mapping of expressions to concrete values

    Returns
    -------
    sqlglot.expressions.Expression
        A sqlglot expression
    """

    gen_alias_index = itertools.count()

    def fn(node, _, **kwargs):
        result = _translate_node(node, **kwargs)

        # don't alias root nodes or value ops
        if node is op or isinstance(node, ops.Value):
            return result

        assert isinstance(node, ops.TableNode)

        alias_index = next(gen_alias_index)
        alias = f"t{alias_index:d}"

        try:
            return result.subquery(alias)
        except AttributeError:
            return sg.alias(result, alias)

    # substitute parameters immediately to avoid having to define a
    # ScalarParameter translation rule
    #
    # this lets us avoid threading `params` through every `translate_val` call
    # only to be used in the one place it would be needed: the ScalarParameter
    # `translate_val` rule
    params = {param.op(): value for param, value in params.items()}
    replace_literals = p.ScalarParameter(dtype=x) >> (
        lambda _, x: ops.Literal(value=params[_], dtype=x)
    )

    # replace the right side of InColumn into a scalar subquery for sql
    # backends
    replace_in_column_with_table_array_view = p.InColumn(..., y) >> _.copy(
        options=c.TableArrayView(
            c.Selection(table=lambda _, y: find_first_base_table(y), selections=(y,))
        ),
    )

    # replace any checks against an empty right side of the IN operation with
    # `False`
    replace_empty_in_values_with_false = p.InValues(..., ()) >> c.Literal(
        False, dtype="bool"
    )

    # subtract one from ranking functions to convert from 1-indexed to 0-indexed
    subtract_one_from_ranking_functions = p.WindowFunction(
        p.RankBase | p.NTile
    ) >> c.Subtract(_, 1)

    add_one_to_nth_value_input = p.NthValue >> _.copy(nth=c.Add(_.nth, 1))

    op = op.replace(
        replace_literals
        | replace_in_column_with_table_array_view
        | replace_empty_in_values_with_false
        | subtract_one_from_ranking_functions
        | add_one_to_nth_value_input
    )
    # apply translate rules in topological order
    node = op.map(fn)[op]
    return node.this if isinstance(node, sg.exp.Subquery) else node
