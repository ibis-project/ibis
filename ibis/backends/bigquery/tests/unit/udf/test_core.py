from __future__ import annotations

import builtins
import sys
import tempfile

import pytest

from ibis.backends.bigquery.udf.core import PythonToJavaScriptTranslator, SymbolTable


def test_symbol_table():
    symbols = SymbolTable()
    assert symbols["a"] == "let a"
    assert symbols["a"] == "a"


def compile(f):
    return PythonToJavaScriptTranslator(f).compile()


def test_function_def(snapshot):
    def f(a, b):
        return a + b

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_variable_declaration(snapshot):
    def f():
        c = 1
        return c + 2

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_yield(snapshot):
    def f(a):
        yield from [1, 2, 3]

    js = compile(f)
    snapshot.assert_match(js, "out.js")


@pytest.mark.skipif(sys.platform == "win32", reason="Skip on Windows")
def test_yield_from(snapshot):
    d = {}

    with tempfile.NamedTemporaryFile("r+") as f:
        f.write(
            """\
def f(a):
    yield from [1, 2, 3]"""
        )
        f.seek(0)
        code = builtins.compile(f.read(), f.name, "exec")
        exec(code, d)
        f = d["f"]
        js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_assign(snapshot):
    def f():
        a = 1
        a = 2
        print(a)  # noqa: T201
        return 1

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def add(x, y):
    return x + y


def sub(x, y):
    return x - y


def mul(x, y):
    return x * y


def div(x, y):
    return x / y


@pytest.mark.parametrize("op", [add, sub, mul, div])
def test_binary_operators(op, snapshot):
    js = compile(op)
    snapshot.assert_match(js, "out.js")


def test_pow(snapshot):
    def f():
        a = 1
        return a**2

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_floordiv(snapshot):
    def f():
        a = 1
        return a // 2

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_unary_minus(snapshot):
    def f():
        a = 1
        return -a

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_unary_plus(snapshot):
    def f():
        a = 1
        return +a

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_true_false_none(snapshot):
    def f():
        a = True
        b = False
        c = None
        return a if c != None else b  # noqa: E711

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_str(snapshot):
    def f():
        a = "abc"
        return a

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_tuple(snapshot):
    def f():
        a = "a", "b", "c"
        return a

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_dict(snapshot):
    def f():
        a = {"a": 1, "b": 2}
        return a

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_continue(snapshot):
    def f():
        i = 0
        for i in [1, 2, 3]:
            if i == 1:
                continue
        return i

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_lambda_with_splat(snapshot):
    def f():
        def sum(sequence):
            total = 0
            for value in sequence:
                total += value
            return total

        splat_sum = lambda *args: sum(args)
        return splat_sum(1, 2, 3)

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_logical_not(snapshot):
    def f():
        a = True
        b = False
        return not a and not b

    js = compile(f)
    snapshot.assert_match(js, "out.js")


class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height


class FancyRectangle(Rectangle):
    @property
    def perimeter(self):
        return self.width * 2 + self.height * 2


def test_class(snapshot):
    js = compile(Rectangle)
    snapshot.assert_match(js, "out.js")


def test_class_with_properties(snapshot):
    js = compile(FancyRectangle)
    snapshot.assert_match(js, "out.js")


def test_set_to_object(snapshot):
    def f(a):
        x = set()
        y = 1
        x.add(y)
        return y

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_setitem(snapshot):
    def f(a):
        x = {}
        y = "2"
        x[y] = y
        return x

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_delete(snapshot):
    def f(a):
        x = [a, 1, 2, 3]
        y = {"a": 1}
        del x[0], x[1]
        del x[0 + 3]
        del y.a
        return 1

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_scope_with_while(snapshot):
    def f():
        class Foo:
            def do_stuff(self):
                while True:
                    i = 1
                    i + 1
                    break

                while True:
                    i = 1
                    i + 1
                    break

                return 1

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_list_comp(snapshot):
    def f():
        x = [a + b for a, b in [(1, 2), (3, 4), (5, 6)] if a > 1 if b > 2]
        return x

    js = compile(f)
    snapshot.assert_match(js, "out.js")


@pytest.mark.xfail(raises=NotImplementedError, reason="Not yet implemented")
def test_nested_list_comp(snapshot):
    # TODO(phillipc): This can be done by nesting map calls followed by
    # N - 1 calls to array.reduce(Array.concat), where N is the number of
    # generators in the comprehension.
    def f():
        x = [
            a + b + c
            for a in [1, 4, 7]
            for b in [2, 5, 8]
            for c in [3, 6, 9]
            if a > 1
            if b > 2
            if c > 3
        ]
        return x

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_splat(snapshot):
    def f(x, y, z):
        def g(a, b, c):
            return a - b - c

        args = [x, y, z]
        return g(*args)

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_varargs(snapshot):
    def f(*args):
        return sum(*args)

    js = compile(f)
    snapshot.assert_match(js, "out.js")


def test_missing_vararg(snapshot):
    def my_range(n):
        return [1 for x in [n]]

    js = compile(my_range)
    snapshot.assert_match(js, "out.js")


def test_len_rewrite(snapshot):
    def my_func(a):
        return len(a)

    js = compile(my_func)
    snapshot.assert_match(js, "out.js")
