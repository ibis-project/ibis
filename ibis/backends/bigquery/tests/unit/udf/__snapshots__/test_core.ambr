# name: test_assign
  '''
  function f() {
      let a = 1;
      a = 2;
      console.log(a);
      return 1;
  }
  '''
# ---
# name: test_binary_operators[mul]
  '''
  function mul(x, y) {
      return (x * y);
  }
  '''
# ---
# name: test_class
  '''
  class Rectangle {
      constructor(width, height) {
          this.width = width;
          this.height = height;
      }
      area() {
          return (this.width * this.height);
      }
  }
  '''
# ---
# name: test_class_with_properties
  '''
  class FancyRectangle extends Rectangle {
      get perimeter() {
          return ((this.width * 2) + (this.height * 2));
      }
  }
  '''
# ---
# name: test_continue
  '''
  function f() {
      let i = 0;
      for (let i of [1, 2, 3]) {
          if ((i === 1)) {
              continue;
          }
      }
      return i;
  }
  '''
# ---
# name: test_delete
  '''
  function f(a) {
      let x = [a, 1, 2, 3];
      let y = {['a']: 1};
      delete x[0];
      delete x[1];
      delete x[(0 + 3)];
      delete y.a;
      return 1;
  }
  '''
# ---
# name: test_dict
  '''
  function f() {
      let a = {['a']: 1, ['b']: 2};
      return a;
  }
  '''
# ---
# name: test_floordiv
  '''
  function f() {
      let a = 1;
      return Math.floor(a / 2);
  }
  '''
# ---
# name: test_function_def
  '''
  function f(a, b) {
      return (a + b);
  }
  '''
# ---
# name: test_lambda_with_splat
  '''
  function f() {
      function sum(sequence) {
          let total = 0;
          for (let value of sequence) {
              total += value;
          }
          return total;
      }
      let splat_sum = ((...args) => sum(args));
      return splat_sum(1, 2, 3);
  }
  '''
# ---
# name: test_len_rewrite
  '''
  function my_func(a) {
      return a.length;
  }
  '''
# ---
# name: test_list_comp
  '''
  function f() {
      let x = [[1, 2], [3, 4], [5, 6]].filter((([a, b]) => ((a > 1) && (b > 2)))).map((([a, b]) => (a + b)));
      return x;
  }
  '''
# ---
# name: test_logical_not
  '''
  function f() {
      let a = true;
      let b = false;
      return ((!a) && (!b));
  }
  '''
# ---
# name: test_missing_vararg
  '''
  function my_range(n) {
      return [n].map(((x) => 1));
  }
  '''
# ---
# name: test_pow
  '''
  function f() {
      let a = 1;
      return Math.pow(a, 2);
  }
  '''
# ---
# name: test_scope_with_while
  '''
  function f() {
      class Foo {
          do_stuff() {
              while (true) {
                  let i = 1;
                  (i + 1);
                  break;
              }
              while (true) {
                  let i = 1;
                  (i + 1);
                  break;
              }
              return 1;
          }
      }
  }
  '''
# ---
# name: test_set_to_object
  '''
  function f(a) {
      let x = (new Set());
      let y = 1;
      x.add(y);
      return y;
  }
  '''
# ---
# name: test_setitem
  '''
  function f(a) {
      let x = {};
      let y = '2';
      x[y] = y;
      return x;
  }
  '''
# ---
# name: test_splat
  '''
  function f(x, y, z) {
      function g(a, b, c) {
          return ((a - b) - c);
      }
      let args = [x, y, z];
      return g(...args);
  }
  '''
# ---
# name: test_str
  '''
  function f() {
      let a = 'abc';
      return a;
  }
  '''
# ---
# name: test_true_false_none
  '''
  function f() {
      let a = true;
      let b = false;
      let c = null;
      return ((c !== null) ? a : b);
  }
  '''
# ---
# name: test_tuple
  '''
  function f() {
      let a = ['a', 'b', 'c'];
      return a;
  }
  '''
# ---
# name: test_unary_minus
  '''
  function f() {
      let a = 1;
      return (-a);
  }
  '''
# ---
# name: test_unary_plus
  '''
  function f() {
      let a = 1;
      return (+a);
  }
  '''
# ---
# name: test_varargs
  '''
  function f(...args) {
      return sum(...args);
  }
  '''
# ---
# name: test_variable_declaration
  '''
  function f() {
      let c = 1;
      return (c + 2);
  }
  '''
# ---
# name: test_yield
  '''
  function* f(a) {
      yield* [1, 2, 3];
  }
  '''
# ---
# name: test_yield_from
  '''
  function* f(a) {
      yield* [1, 2, 3];
  }
  '''
# ---
